File: 3532.path-existence-queries-in-a-graph-i.cpp
// @leet start
#include <numeric>
#include <vector>
using namespace std;
class Solution {
    struct UnionFind {
        vector<int> parent;

        UnionFind(int n) : parent(n) { iota(parent.begin(), parent.end(), 0); }

        int find(int x) {
            if (parent[x] == x) return x;
            // no need to compress
            // because the depth of x won't be more than 2
            return find(parent[x]);
        }

        void unite(int x, int y) {
            int root_x = find(x), root_y = find(y);
            if (root_x != root_y) { parent[root_y] = parent[root_x]; }
        }

        bool is_connected(int x, int y) { return find(x) == find(y); }
    };

public:
    vector<bool> pathExistenceQueries(int n, vector<int>& nums, int maxDiff,
                                      vector<vector<int>>& queries) {
        // diff(nums[i], nums[j]) <= maxDiff -> i <--> j
        UnionFind uf(n);
        int num_queries = queries.size();
        vector<bool> results;
        results.reserve(num_queries);
        for (int i = 0; i < n - 1; ++i) {
            if (nums[i + 1] - nums[i] <= maxDiff) { uf.unite(i, i + 1); }
        }

        for (auto& query : queries) {
            if (uf.is_connected(query[0], query[1])) {
                results.push_back(true);
            } else {
                results.push_back(false);
            }
        }

        return results;
    }
};
// @leet end

File: 3532.path-existence-queries-in-a-graph-i.rs
// @leet start
struct UnionFind {
    parent: Vec<usize>,
}

impl UnionFind {
    fn new(n: usize) -> Self {
        UnionFind {
            parent: (0..n).collect(),
        }
    }

    fn find(&self, x: usize) -> usize {
        if self.parent[x] == x {
            return x;
        }

        self.find(self.parent[x])
    }

    fn unite(&mut self, x: usize, y: usize) {
        let px = self.find(x);
        let py = self.find(y);

        if px == py {
            return;
        }

        self.parent[py] = px;
    }

    fn is_connected(&self, x: usize, y: usize) -> bool {
        self.find(x) == self.find(y)
    }
}

impl Solution {
    pub fn path_existence_queries(
        n: i32,
        nums: Vec<i32>,
        max_diff: i32,
        queries: Vec<Vec<i32>>,
    ) -> Vec<bool> {
        let n = n as usize;
        let mut uf = UnionFind::new(n);
        for i in 0..n - 1 as usize {
            if nums[i + 1] - nums[i] <= max_diff {
                uf.unite(i, i + 1);
            }
        }

        let mut result = Vec::with_capacity(queries.len());
        for query in queries {
            if uf.is_connected(query[0] as usize, query[1] as usize) {
                result.push(true);
            } else {
                result.push(false);
            }
        }

        result
    }
}
// @leet end
